openhands/llm/fn_call_converter.py

```diff
@@ -0,0 +1,795 @@
+"""Convert function calling messages to non-function calling messages and vice versa.
+
+This will inject prompts so that models that doesn't support function calling
+can still be used with function calling agents.
+
+We follow format from: https://docs.litellm.ai/docs/completion/function_call
+"""
+
+import copy
+import json
+import re
+from typing import Iterable
+
+from litellm import ChatCompletionToolParam
+
+from openhands.core.exceptions import (
+    FunctionCallConversionError,
+    FunctionCallValidationError,
+)
+
+# Inspired by: https://docs.together.ai/docs/llama-3-function-calling#function-calling-w-llama-31-70b
+SYSTEM_PROMPT_SUFFIX_TEMPLATE = """
+You have access to the following functions:
+
+{description}
+
+If you choose to call a function ONLY reply in the following format with NO suffix:
+
+<function=example_function_name>
+<parameter=example_parameter_1>value_1</parameter>
+<parameter=example_parameter_2>
+This is the value for the second parameter
+that can span
+multiple lines
+</parameter>
+</function>
+
+<IMPORTANT>
+Reminder:
+- Function calls MUST follow the specified format, start with <function= and end with </function>
+- Required parameters MUST be specified
+- Only call one function at a time
+- You may provide optional reasoning for your function call in natural language BEFORE the function call, but NOT after.
```

Comment:
This sounds reasonable, but we could also be lenient with it? Like we could discard whatever follows /function.

Reply:
yeah - we set stop word to `</function`